Update mutation checking in pattern matcher

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137449
* __->__ #137448

Fix for https://github.com/pytorch/pytorch/issues/137229

The current mutation checking is complicated because it works for pre grad IR. When pre grad ir has been traced to OpOverloads checking is much easier. I am also special casing auto functional hop although I discussed with @zou3519 it would be nice to have a way of querying HOPs that mimic schemas.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

eellison committed:
Update mutation checking in pattern matcher

[ghstack-poisoned]

eellison committed:
Update on "Update mutation checking in pattern matcher"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

zou3519 commented on torch/_inductor/config.py:
```diff
@@ -39,7 +39,7 @@ def autotune_remote_cache_default() -> Optional[bool]:
 
 # use fx aot graph codegen cache
 fx_graph_cache = (
-    os.environ.get("TORCHINDUCTOR_FX_GRAPH_CACHE", "0" if is_fbcode() else "1") == "1"
```
Is this intentional?

eellison replied:
noo

zou3519 commented on torch/_inductor/pattern_matcher.py:
```diff
@@ -1616,7 +1616,26 @@ def is_start_of_fx_graph(graph: torch.fx.Graph, node: torch.fx.Node) -> bool:
 _mutation_op_re = re.compile(r"(?<!_)(_$|_[.]|(\b|_)(set|enter|exit|seed)(\b|_))(?!_)")
 
 
+def incorrect_inductor_schema_op(op: torch._ops.OpOverload) -> bool:
```
nit: FIXME in function name?, with note in the TODO to not add anything here

eellison replied:
I have a pr after this stack that fixes them, but not sure if there are any interactions with functionalization that adding the mutation breaks cc @bdhirsh

eellison replied:
FWIW, i did open a pr updating, and things break... leaving as it is for now

eellison committed:
Update on "Update mutation checking in pattern matcher"


Fix for https://github.com/pytorch/pytorch/issues/137229

The current mutation checking is complicated because it works for pre grad IR. When pre grad ir has been traced to OpOverloads checking is much easier. I am also special casing auto functional hop although I discussed with zou3519 it would be nice to have a way of querying HOPs that mimic schemas.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge
